src/community/mapml/src/main/java/org/geoserver/mapml/tcrs/TiledCRS.java

```diff
@@ -0,0 +1,284 @@
+        Point max = this.transformation.transform(bounds.max, this.scales[zoom]).round();
+        return new Bounds(min, max);
+    }
+    // convenience methods
+    public Point project(LatLng latLng) {
+        return this.projection.project(latLng);
+    }
+
+    public LatLng unproject(Point point) {
+        return this.projection.unproject(point);
+    }
+    /**
+     * Count the width of the *tile* bounds at the given zoom level in integral tile units.
+     *
+     * @param zoom integer zoom level at which to calculate the width
+     * @param bounds the bounds for which the calculation/conversion should be done
+     * @return long the number of tiles wide the bounds is at the zoom level
+     */
+    protected long tileWidth(int zoom, Bounds bounds) {
+        if (zoom == -1 || bounds == null) return 0;
+        return (long) bounds.max.x + 1 - (long) bounds.min.x;
+    }
+
+    public long tileCount(int zoom, Bounds bounds) {
+        if (zoom == -1 || bounds == null) return 0;
+
+        Point min = bounds.min.divideBy(TILESIZE).floor();
+        Point max = bounds.max.divideBy(TILESIZE).floor();
+
+        // integer coordinate system, bump max values up to next increment
+        long width = (long) (max.x + 1 - min.x);
+        long height = (long) (max.y + 1 - min.y);
+        return width * height;
+    }
+
+    public List<TileCoordinates> getTilesForExtent(Bounds extent, int zoom, long start) {
+
+        // the extent must be expressed in projected, scaled units
+        Bounds pb = extent;
+        // the min/max in decimal tiles truncated to the next lower integer tile ordinate
+        Bounds tb =
+                new Bounds(pb.min.divideBy(TILESIZE).floor(), pb.max.divideBy(TILESIZE).floor());
+        long width = tileWidth(zoom, tb);
+        List<TileCoordinates> tiles = new ArrayList<>();
+        for (long i = (start > 0 ? (long) tb.min.y + start / width : (long) tb.min.y);
+                i <= tb.max.y;
+                i++) {
+            for (long j = (start > 0 ? (long) tb.min.x + (start % width) : (long) tb.min.x);
+                    j <= tb.max.x;
+                    j++) {
+                if (tiles.size() < pageSize) {
+                    if (i >= 0
+                            && i <= tileBounds.get(zoom).max.y
+                            && j >= 0
+                            && j < tileBounds.get(zoom).max.x) {
+                        tiles.add(new TileCoordinates(j, i, zoom));
+                    }
+                } else {
+                    break;
+                }
+            }
+        }
+        // the centre of the extent in decimal tiles... not truncated
+        Point centre = pb.getCentre().divideBy(TILESIZE);
+        Collections.sort(tiles, new TileComparator(centre));
+        return tiles;
+    }
+
+    public List<TileCoordinates> getTilesForExtent(LatLngBounds extent, int zoom, long start) {
```

Comment:
If it helps geotools has a WMTS client extension that has very similar logic to what you are accomplishing here.

Reply:
I'll have a look, thanks!